### PR TITLE
[WIP] Overlays

### PR DESCRIPTION
**Pull request by brisvag**
# Description
Here's a [very work-in-progress] attempt at unifying overlays (see #3755 ).

Overlays are currently all broken (and a lot of the diff is for now simply removing stuff that would otherwise crash napari), except for `Text`, which is now using the new system. I thought this was a good point to create a PR and ask for feedback!

EDIT: current state:
- [x] text
- [x] scale_bar
- [ ] axis (should be here?)
- [ ] interaction box 

### Goals

- move overlays to an `Overlays` evented model on the viewer
- generate overlays similarly to how layers are generated (automatically from the model)
- use them for everything (mainly selection boxes)

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/_qt/qt_viewer.py [hunk 1]
```diff
@@ -380,34 +380,49 @@ def _create_canvas(self) -> None:
...
+        for i, (name, overlay) in enumerate(self.viewer.overlays):
+            if name == 'visible':
+                continue
+            vispy_overlay = create_vispy_overlay(
+                overlay, viewer=self.viewer, parent=self.view, order=1e6 + i
+            )
+
+            # we must keep a reference of they will disappear
+            # TODO: using name here because evented models are unhashable?
+            self.overlay_to_visual[name] = vispy_overlay
```
Comment (brisvag) [inline napari/_qt/qt_viewer.py]: Relevant changes in `qt_viewer` are here.

File: napari/_vispy/overlays/base.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        self.node.visible = (
+            self.viewer.overlays.visible and self.overlay.visible
+        )
+
+    def _on_position_change(self, event=None):
+        pass
+
+    def reset(self):
+        self._on_visible_change()
+        self._on_position_change()
```
Comment (brisvag) [inline napari/_vispy/overlays/base.py]: This is the overlay equivalent of `VispyBaseLayer`.

File: napari/_vispy/overlays/text.py [hunk 3]
```diff
@@ -1,80 +1,66 @@
...
-from vispy.visuals.transforms import STTransform
 
-from ...components._viewer_constants import TextOverlayPosition
+from ...components._viewer_constants import Position
 from ...utils.translations import trans
+from .base import VispyBaseOverlay
 
 
-class VispyTextOverlay:
+class VispyTextOverlay(VispyBaseOverlay):
```
Comment (brisvag) [inline napari/_vispy/overlays/text.py:8]: The only overlay currently functioning is this!

File: napari/_vispy/utils/visual.py [hunk 4]
```diff
@@ -70,6 +84,37 @@ def create_vispy_visual(layer: Layer) -> VispyBaseLayer:
...
+        if isinstance(overlay, overlay_type):
+            return visual_class(overlay=overlay, **kwargs)
+
+    raise TypeError(
+        trans._(
+            'Could not find VispyOverlay for overlay of type {dtype}',
+            deferred=True,
+            dtype=type(overlay),
+        )
+    )
```
Comment (brisvag) [inline napari/_vispy/utils/visual.py:116]: This is copyign the implementation of `create_vispy_layer`.

File: napari/components/overlays/overlays.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    # default_factory=InteractionBox, allow_mutation=False
+    # )
+    # axes: Axes = Field(
+    # default_factory=Axes, allow_mutation=False
+    # )
+    text: Text = Field(default_factory=Text, allow_mutation=False)
+    # scale_bar: ScaleBar = Field(
+    # default_factory=ScaleBar, allow_mutation=False
+    # )
+    visible: bool = True
```
Comment (brisvag) [inline napari/components/overlays/overlays.py]: This is the core evented model. currently, order of fields also controls drawing order in vispy, but that's probably bad.

**Review by brisvag — COMMENTED** [review]
I left a few comments to point out the important changes. Please ignore any changes specific to other non-text overlays: they are temporary, and broken still.

**Comment by alisterburt** [comment]
looks cool! I see that the interaction box and axis visual are both overlays here, does this mean you've moved away from your strict definition of overlays as things which exist in the canvas coordinate system or is it just temporary?

**Comment by brisvag** [comment]
> looks cool! I see that the interaction box and axis visual are both overlays here, does this mean you've moved away from your strict definition of overlays as things which exist in the canvas coordinate system or is it just temporary?

They do exist only in the canvas coordinate system. Are you referring to the fact that Axis is affected by the camera orientation? In my mind, that does not matter. What (I think) should differentiate layers and overlays is:

- layers contain Data
- layers appear in the layer list
- the contents of layers *live* in world space (i.e: they use an `Affine` transform) while overlays live in the canvas (i.e: they use `STTransform`)

EDIT: I just now started playign aroudn with axes (which I had never really sued before) and I see what you mean now. I assumed the axes were displayed in the corner of the canvas (much like many vispy examples do), not in the corner of the world space. Not sure how to taclke that, then. As is, it should probably NOT be an overlay, but one of your (@alisterburt, @kevinyamauchi  ) 3D objects.

**Comment by brisvag** [comment]
Interaction box is now partially working, in the sense that it's attached to the canvas and connected to the model via events.

What's missing is the connection to the mouse bindings, which appears to be broken by something in my chnages. @jojoelfe , could you check it out and see if you spot the issues?

To test out that the evented model works, and it's correctly connected to the visual:

```python
import napari
import numpy as np

v = napari.Viewer()
il = v.add_image(np.random.rand(10, 10, 10))
v.overlays.interaction_box.update(dict(
    visible=True,
    points=np.array([[1, 1], [1, 2], [2, 1], [2, 2]]) * 300,
    show_handle=True,
    show_vertices=True,
))
```

however, you'll see that as soon as you click somewhere, it disappears. If you enable the transform mode, even stranger things happen (the box appears in the corner, tiny).

```python
il.mode = 'transform'
```

**Comment by brisvag** [comment]
On closer inspection, the interaction box mouse bindings are working correctly (i.e: you can drag the corner of an image just fine!); it's the visualization that's broken and somehow ends up squished in the corner...

File: napari/_qt/qt_viewer.py [hunk 6]
```diff
@@ -370,33 +371,22 @@ def _create_canvas(self) -> None:
...
-    def _add_visuals(self) -> None:
-        """Add visuals for axes, scale bar, and welcome text."""
+    def _add_overlay_visuals(self) -> None:
+        """Add visuals for overlays."""
+
+        for i, (name, overlay) in enumerate(self.viewer.overlays):
+            if name == 'visible':
+                continue
+            vispy_overlay = create_vispy_overlay(
+                overlay, viewer=self.viewer, parent=self.view, order=1e6 + i
```
Comment (jojoelfe) [inline napari/_qt/qt_viewer.py:381]: I think the problem might be that for the interaction_box the parent should be view.scene and not view. I guess that means it wants to live in world coordinates and not in canvas coordinates.

**Comment by jojoelfe** [comment]
I think when I was planning the interaction_box the plan was to have it in canvas coordinates, but in practice it was much easier to have it in world coordinates, because there are `data_to_world` functions, but no easy to access `data_to_canvas` or `world_to_canvas` functions.  #3783 should help there.

The decision was also easy because the main practical reason to have it in canvas coordinates is that control-point size should be independent of camera zoom, but you get that behavior for free from `.scaling = False` of the marker visual.

**Comment by brisvag** [comment]
Oh, I didntt realize! Yeah that makes sense. However, it seems like we should definitely move towards canvas coordinates for something like this...

**Comment by sofroniewn** [comment]
hi @brisvag what are your thoughts on picking this up again? i can start giving it some review. In general though, I think the addition of `VispyBaseOverlay` and `Overlays` will be good for the code base. This is also something that @andy-sweet might be able to help review too

**Comment by brisvag** [comment]
I'm happy to pick it up again, but we need some discussion before we can move forward I think. As you can see from the comments above, there are a few things that make the transition harder than I thought. Specifically, the axis and interaction box are currently tied to the world coordinates, which makes them unsuitable for the `Overlay` abstraction.

While I'd be ok with leaving `Axes` out of this if most people prefer the current behaviour (that is, they are fixed in a point in world coordinates rather than being in a corner of the canvas), I think we really need to figure out how to get `InteractionBox` "fixed" to work as an overlay. It seems that this requires some refactoring on the transform logic, for which I'll definitely need help, maybe from @andy-sweet !

**Comment by Carreau** [comment]
This PR have not been active for quite some time. 

In order to keep the list of PRs short and easier to work with, I am going to tag this PR as [stale](https://github.com/napari/napari/pulls?q=is%3Apr+label%3Astale+is%3Aclosed) and close it.

Feel free to reopen and remove the "stale" label when work on this PR resumes. 
If you do not have the permissions to do so, feel free to request help from one of the maintainer to reopen and remove the `stale` label .

**Pull request closed without merging**